rcl/src/rcl/node_type_cache.c

```diff
@@ -0,0 +1,254 @@
+  RCL_CHECK_ARGUMENT_FOR_NULL(type_description, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_FOR_NULL_WITH_MSG(node->impl->registered_types_by_type_hash,
+                              "type cache not initialized",
+                              return RCL_RET_NOT_INIT);
+
+  // Convert hash to string
+  if (RCUTILS_RET_OK !=
+      rosidl_stringify_type_hash(type_hash, rcutils_get_default_allocator(),
+                                 &type_hash_str)) {
+    RCL_SET_ERROR_MSG("Failed to stringify type hash");
+    return RCL_RET_ERROR;
+  }
+
+  // TODO(achim-k): Remove log statement
+  const char *typeName = type_description->type_description.type_name.data;
+  RCUTILS_LOG_INFO_NAMED(ROS_PACKAGE_NAME, "R: %s (%s)", type_hash_str,
+                         typeName);
+
+  // If the type already exists, we only have to increment the registration
+  // count.
+  if (rcutils_hash_map_key_exists(node->impl->registered_types_by_type_hash,
+                                  &type_hash_str)) {
+    if (RCUTILS_RET_OK !=
+        rcutils_hash_map_get(node->impl->registered_types_by_type_hash,
+                             &type_hash_str, &type_info)) {
+      RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(
+          "Failed to retrieve type info for type '%s'", type_hash_str);
+      return RCL_RET_ERROR;
+    }
+
+    type_info.numRegistrations++;
+  } else {
+    // First registration of this type
+    type_info.numRegistrations = 1;
+
+    // Convert type description struct to type description message struct.
+    type_info.type_description =
+        rcl_convert_type_description_runtime_to_msg(type_description);
+    RCL_CHECK_FOR_NULL_WITH_MSG(type_info.type_description,
+                                "converting type description struct failed",
+                                return RCL_RET_ERROR);
+  }
+
+  // Update the hash map entry.
+  if (RCUTILS_RET_OK !=
+      rcutils_hash_map_set(node->impl->registered_types_by_type_hash,
+                           &type_hash_str, &type_info)) {
+    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(
+        "Failed to update type info for type '%s'", type_hash_str);
+    return RCL_RET_ERROR;
+  }
+
+  return RCL_RET_OK;
+}
+
+rcl_ret_t rcl_node_type_cache_init(const rcl_node_t *node) {
+  RCL_CHECK_ARGUMENT_FOR_NULL(node, RCL_RET_INVALID_ARGUMENT);
+
+  if (NULL != node->impl->registered_types_by_type_hash) {
+    RCL_SET_ERROR_MSG("Type cache already initialized");
+    return RCL_RET_ALREADY_INIT;
+  }
+
+  node->impl->registered_types_by_type_hash =
+      node->context->impl->allocator.allocate(
+          sizeof(rcutils_hash_map_t), node->context->impl->allocator.state);
+  RCL_CHECK_FOR_NULL_WITH_MSG(node->impl->registered_types_by_type_hash,
+                              "allocating memory failed",
+                              return RCL_RET_BAD_ALLOC);
+
+  *node->impl->registered_types_by_type_hash =
+      rcutils_get_zero_initialized_hash_map();
+
+  rcutils_ret_t ret = rcutils_hash_map_init(
```

Comment:
For performance and simplicity - I believe that the SHA256 value of the hash can already be treated as essentially random. Instead of taking the time/memory to stringify the value and run hash algorithm on the string, it should be a lot cheaper to use the first `sizeof(size_t)` bytes of `type_hash->value` as the hash key. So, something like:

```
size_t get_type_hash_hashmap_key(const rosidl_type_hash_t * type_hash) {
  // Reinterpret-cast the first sizeof(size_t) bytes of the hash value
  return (size_t)*type_hash->value;
}

int cmp_type_hash(const rosidl_type_hash_t * val1, const rosidl_type_hash_t * val2) {
  return memcmp(val1, val2, sizeof(rosidl_type_hash_t));
}

rcutils_hash_map_init(
  hash_map, 0, 
  sizeof(rosidl_type_hash_t), 
  sizeof(rcl_type_info_t), 
  &get_type_hash_hashmap_key, 
  &cmp_type_hash, 
  allocator);
```